### DISCUSSION TITLE
Advanced search in Homer app?

### DISCUSSION THREAD

**Discussion by pekkaar** [sipcapture/homer#532] [Q&A] [answered]
Hello Team!

I was wondering if there is any documentation or tutorial available for searching on the Homer app? The interface is intuitive enough for the basics, but it does not help figuring out advanced stuff, like:

- **how to supply negated values?** (where fields do NOT equal to a specific value)
- **how to perform wildcard searches?** (where I provide some values for the fields and explicitly mark characters I want the search to disregard)
- **how to search for SIP methods that has or does not have a particular SIP header field?** (regardless of whether the header field has any value at all or not -- for example I would like to find all INVITES with "X-CID" header in them, but I don't care if there is no value in this header)
- **is regex supported in any way?**

**Comment by adubovikov** [comment, marked as answer]
Hi,

There is no single "advanced search tutorial" doc, but here is how Homer handles the cases you listed.

### Homer 7 (homer-app + homer-ui)

**Wildcards / partial match**

Many search fields use partial matching when the field mapping has `"index": "wildcard"` (Call-ID, from/to user, message/payload, etc.). You usually type part of the value; the backend adds `LIKE '%value%'`.

**Negated values (NOT equal)**

Not exposed as a general UI operator in the standard search form. Workarounds:

- narrow with other fields (method, time range, IP, user), or
- use payload/message search creatively, or
- query the API/SQL layer if you run a custom integration.

**SIP header present / absent (e.g. INVITE with `X-CID`)**

Standard UI search is column-based, not full SIP header grammar. Practical options:

- search **payload/message** for the header name, e.g. `X-CID` (finds messages containing that header line), or
- add/customize **mapping_schema** fields if your pipeline extracts headers into searchable columns.

There is no built-in checkbox "header exists regardless of value" in Homer 7 out of the box.

**Regex**

Structured search does **not** expose general regex on all fields. Payload/message `LIKE` search is substring-based, not full regex.

---

### Homer 11

Homer 11 improves this, but the same limits apply to the basic form:

| Need | Homer 11 |
|------|------------|
| Partial match | default on Call-ID, caller/callee (`LIKE`) |
| Wildcard in payload | **Payload** filter: `payload LIKE '%text%'` (use `%` yourself) |
| Header present | **Payload** search for `X-CID:` or custom **mapping virtual field** with `present` / `absent` (Settings → Mappings) |
| Negation / regex | not in the simple search widget; use **SQL search** (`POST /api/v4/query`), **`homer search` CLI**, or Search widget **AI/SQL mode** (MCP) |

Example for "INVITEs containing X-CID header":

- Method: `INVITE`
- Payload: `X-CID`

Or in Homer 11 CLI:

```bash
homer search --method INVITE --grep 'X-CID'
```

### Docs worth reading (H11)

- `docs/SEARCH.md` — CLI filters, `--grep` / `--exclude`
- `docs/MCP_UI_GUIDE.md` — natural language / SQL search in the UI
- Settings → **Mappings** — custom virtual fields (present/absent/like on `data_extra`)

**Bottom line:** substring/wildcard-style search works on many fields; negation and full regex are power-user/API/SQL features, not the default Homer 7 form. Header presence is usually done via payload search or custom mapped fields.

If you say whether you are on Homer 7 or 11, we can suggest a concrete mapping or query for your `X-CID` example.
